### PR TITLE
windows: show progress bar in task bar

### PR DESCRIPTION
**Pull request by alexeyneu**
progress bar over  task bar thumbnail button  
Windows only  
untitled 1

### CODE REVIEW

**Comment by laanwj** [comment]
> 120 bucks
> (address removed)

thank you for your first time contribution!

however, it is not-done to solicit donations on github, so I've had to remove the address

File: src/qt/modaloverlay.cpp
Comment (scravy) [inline src/qt/modaloverlay.cpp]: you added a fair bit of trailing white space here ;-)

**Comment by alexeyneu** [comment]
linux `#ifdef` 's were missing. it's reloaded some time ago.  
@laanwj ok ,now you can copy-paste my wallet address by double click , line near location , github profile. I bet you wanna to

File: src/qt/modaloverlay.cpp
Comment (scravy) [reply src/qt/modaloverlay.cpp]: this is actually failing your travis build in the lint stage – `This diff appears to have added new lines with trailing whitespace.` – https://travis-ci.org/bitcoin/bitcoin/jobs/421514716#L719

Comment (alexeyneu) [reply src/qt/modaloverlay.cpp]: should work if it were only unwanted thing

File: src/qt/winshutdownmonitor.cpp
Comment (MarcoFalke) [inline src/qt/winshutdownmonitor.cpp]: Don't forget to install `clang-format` and run the https://github.com/bitcoin/bitcoin/tree/master/contrib/devtools#clang-format-diffpy script to preempt whitespace nitpicking.

File: src/qt/winshutdownmonitor.cpp
Comment (donaloconnor) [inline src/qt/winshutdownmonitor.cpp]: What's the purpose of com_b here? This code is very confusing

File: src/qt/winshutdownmonitor.cpp
Comment (donaloconnor) [inline src/qt/winshutdownmonitor.cpp]: `IID_ITaskbarList3` is only available on Windows 7+. I think Bitcoin Core supports Win Vista+. This will probably fail to initialise and return a null bhr. We should ensure this succeeds.

File: src/qt/winshutdownmonitor.cpp
Comment (donaloconnor) [inline src/qt/winshutdownmonitor.cpp]: 184 magic number?

File: src/qt/modaloverlay.cpp
Comment (donaloconnor) [inline src/qt/modaloverlay.cpp]: What's the 190?

**Review by donaloconnor — CHANGES_REQUESTED** [review]
Concept ACK but...
I find this code very difficult to follow in its current form. There are a lot of magic numbers, variable names that are confusing etc.

Also should we consider using QWinTaskbarProgress ?

File: src/qt/winshutdownmonitor.cpp
Comment (alexeyneu) [reply src/qt/winshutdownmonitor.cpp]: fine to read .   Just checked petzold , yeah  `return` end is the common practice

File: src/qt/winshutdownmonitor.cpp
Comment (alexeyneu) [reply src/qt/winshutdownmonitor.cpp]: i'm not sure about vista also.

File: src/qt/winshutdownmonitor.cpp
Comment (alexeyneu) [reply src/qt/winshutdownmonitor.cpp]: 2.5% left . Or you'll have green shade all the time.

File: src/qt/modaloverlay.cpp
Comment (alexeyneu) [reply src/qt/modaloverlay.cpp]: size of button on 4k i bet

File: src/qt/winshutdownmonitor.cpp
Comment (alexeyneu) [reply src/qt/winshutdownmonitor.cpp]: to show that first two windows are not used. i'm not sure `++com_b > 2` will be  a lot better. short-circuit  evaluation itself is not a problem i guess.

File: src/qt/modaloverlay.cpp
Comment (alexeyneu) [reply src/qt/modaloverlay.cpp]: about QWinTaskbarProgress    something tells me you have reasons not to use it haha

Comment (scravy) [reply src/qt/modaloverlay.cpp]: the button as in the application icon in the task bar? In what unit, pixels? And if so, will that width change if you for instance put your task bar on the left or on the right (where they look completely different and are as horizontally spacious as you want).

The only problem I can see with QWinTaskbarProgress is that is seems to be there only since Qt 5.2 and it's part of winextras.

Comment (donaloconnor) [reply src/qt/modaloverlay.cpp]: Also if button sizes are taken into consideration, then DPI scaling should also be considered

Comment (alexeyneu) [reply src/qt/modaloverlay.cpp]: it's never change You could put 1190 with np . If you put say 30 .... I bet you'd understand already

Comment (alexeyneu) [reply src/qt/modaloverlay.cpp]: yeah that's about hsize in pixels. 
@scravy about latter , trust me it's not only problem

Comment (alexeyneu) [reply src/qt/modaloverlay.cpp]: this square with bitcoin logo in taskbar called 'button' in related docs

File: src/qt/winshutdownmonitor.cpp
Comment (alexeyneu) [reply src/qt/winshutdownmonitor.cpp]: looks like travis is stuck. there's `\nLF` there.

File: src/qt/bitcoin.cpp
Comment (luke-jr) [inline src/qt/bitcoin.cpp]: Please don't remove unrelated whitespace.

Comment (alexeyneu) [reply src/qt/bitcoin.cpp]: that's not funny

Comment (luke-jr) [reply src/qt/bitcoin.cpp]: It wasn't meant to be funny...?

Comment (promag) [reply src/qt/bitcoin.cpp]: @alexeyneu you should only commit what is relevant to the PR.

File: src/qt/winshutdownmonitor.cpp
Comment (scravy) [reply src/qt/winshutdownmonitor.cpp]: @alexeyneu this is *not* resolved. The `break` after a return is dead code and should be removed.

File: src/qt/winshutdownmonitor.cpp
Comment (scravy) [inline src/qt/winshutdownmonitor.cpp]: same here

File: src/qt/winshutdownmonitor.cpp
Comment (scravy) [reply src/qt/winshutdownmonitor.cpp]: To make this apparently magic number more self describing, maybe extract it into a constant that has an explanatory name or add a comment describing why 184.

File: src/qt/modaloverlay.h
Comment (scravy) [inline src/qt/modaloverlay.h]: it is unnecessary to replace the space indentation with tabs indentation, please don't.

File: src/qt/bitcoin.cpp
Comment (alexeyneu) [reply src/qt/bitcoin.cpp]: who you mean by developer?And what about  this pair of your tools , does its build results matter? And one of you two for sure should know do you provide vista support?

Comment (scravy) [reply src/qt/bitcoin.cpp]: @alexeyneu https://github.com/bitcoin/bitcoin/blob/master/doc/developer-notes.md

Comment (scravy) [reply src/qt/bitcoin.cpp]: This document gives a clue about the expected compatibility: https://github.com/bitcoin/bitcoin/blob/master/doc/release-notes.md#compatibility

File: src/qt/winshutdownmonitor.cpp
Comment (alexeyneu) [reply src/qt/winshutdownmonitor.cpp]: for now i wait successful build. Does it worth it?

File: src/qt/bitcoin.cpp
Comment (alexeyneu) [reply src/qt/bitcoin.cpp]: so no vista

**Comment by promag** [comment]
Concept ACK and agree with @donaloconnor. The implementation should be should be decoupled as much as possible.

> Also should we consider using QWinTaskbarProgress ?

That requires another dependency `QT += winextras` (?) and Qt 5.2.

File: src/qt/winshutdownmonitor.cpp
Comment (alexeyneu) [reply src/qt/winshutdownmonitor.cpp]: @scravy   Did you look to left side of diff?

**Comment by promag** [comment]
After seeing [QWinTaskbarProgress](http://doc.qt.io/qt-5/qwintaskbarprogress.html) I think we should use it instead of adding more OS specific code — that's what Qt is for?

**Comment by jonasschnelli** [comment]
Agree with @promag, using `QWinTaskbarProgress` makes far more sense to me.

File: src/qt/winshutdownmonitor.cpp
Comment (alexeyneu) [reply src/qt/winshutdownmonitor.cpp]: and you'd stay silent all this time? how it comes

**Comment by luke-jr** [comment]
@alexeyneu What is your native language?

Can you do this using `QWinTaskbarProgress`?

**Comment by alexeyneu** [comment]
Russian .  
No i wont help you on that.
Unrelated to that: my opinion  - it's a bad idea

File: src/qt/winshutdownmonitor.cpp
Comment (alexeyneu) [reply src/qt/winshutdownmonitor.cpp]: It's here for a reason .You need either left one or mine . Otherwise build here starts to fail.

Comment (scravy) [reply src/qt/winshutdownmonitor.cpp]: screen shot 2018-08-31 at 18 09 11

Yes, what about it?

Comment (scravy) [reply src/qt/winshutdownmonitor.cpp]: long story short: There is no actual change here. Just don't touch these lines and it's less code to discuss, reason, build-error about.

Comment (alexeyneu) [reply src/qt/winshutdownmonitor.cpp]: so if i marked it as resolved then it is

**Comment by alexeyneu** [comment]
it's ready

**Comment by fanquake** [comment]
I've updated the PR title and removed unrelated text from the description.

@alexeyneu This is not ready for merge yet. Please update your commit with a more meaningful commit message, and at least fix any linting issues, see [here](https://travis-ci.org/bitcoin/bitcoin/jobs/423500760).

However I also agree with the commenters above that using `QWinTaskbarProgress` would seem like a better idea than the current implementation. So I don't think this will be merged in it's current state. Can you also justify not using it with more information than:
> Unrelated to that: my opinion - it's a bad idea

**Comment by alexeyneu** [comment]
@fanquake  check chat before writing this bot-like stuff  
https://botbot.me/freenode/bitcoin-core-dev/2018-09-01/?msg=104023171&page=2

**Comment by scravy** [comment]
@alexeyneu I don't know what your point is in linking this chatlog. Are you referring to the build which has no errors but the other stages passed?

If you compare that build with your build you will notice that in your build the checks on different platforms have not even been executed. This is because the lint stage failed and it failed correctly. You are messing up white space for no reason and this has been pointed out several times in this pull request already, you are just choosing to neglect these comments.

**Comment by alexeyneu** [comment]
@scavy dont show up here any more

**Comment by fanquake** [comment]
@alexeyneu Given that you've chosen to ignore review comments and constructive criticism, asked for donations [multiple](https://github.com/bitcoin/bitcoin/pull/14090#issue-211373731) [times](https://botbot.me/freenode/bitcoin-core-dev/2018-09-02/?msg=104043548&page=1), [called people "morons"](https://botbot.me/freenode/bitcoin-core-dev/2018-09-02/?msg=104044512&page=1) and told others not to ["show up here any more"](https://github.com/bitcoin/bitcoin/pull/14090#issuecomment-417929318), I'm inclined to close this pull request.

I appreciate there is a language barrier, and this is your first time contributing (which can be tough), so if you want to keep working on this PR, and discuss, in detail, the downsides to using `QWinTaskbarProgress`, lets do that. 

However, keep the comments constructive and on topic, otherwise this PR will be closed.

**Comment by sipa** [comment]
@alexeyneu Just to be clear: regarding your earlier comment on IRC, the Bitcoin Core project does not have funds for paying for contributions. If your patch has a price, please close it.

There are multiple ways of being funded to contribute to open source projects, but posting a BTC address and putting a price on a PR is not one of them. You could find someone to sponsor your time, or otherwise employ you to contribute.

Also, you're testing people's patience for all the reasons like by @fanquake above. We're happy for your contributions, but if you want to see your patch accepted, please don't dismiss review comments from maintainers and others, and be respectful.

**Comment by promag** [comment]
Thanks, but no thanks. NACK current implementation and IMO not worth it to have such a specific OS feature unless abstracted by Qt.

**Comment by sipa** [comment]
In that case, we have no interest in your code.

**Comment by sipa** [comment]
In that case, we have no interest in your code.

**Comment by donaloconnor** [comment]
NACK. 

I helped review and follow the discussions in good faith. I believed it was a language barrior. Plenty of others would be more than happy to contribute something like this and be nice and unconditional about it.

**Pull request closed without merging**
